### PR TITLE
Pm new get

### PR DESCRIPTION
**Pull request by 70ray**
Changed back to GET instead of POST for project search so that searches can be bookmarked.
Add an invert filter checkbox for the language selector so that it is possible to search for all languages except a few. Otherwise selecting most of the language options would make the url too long. The other multi-selects do not have this problem because there are fewer options. This also simplifies the code slightly.

### CODE REVIEW

File: pinc/ProjectSearchResults.inc [hunk 1]
```diff
@@ -107,9 +107,13 @@ class Column
                 if($sec_sort == $this->id)
                     $sec_sorter = "&nbsp;^";
```
Comment (cpeel) [inline pinc/ProjectSearchResults.inc]: If you're going to use brackets for the `else`, use them for the `if` part too.

File: pinc/ProjectSearchResults.inc [hunk 2]
```diff
@@ -650,9 +641,21 @@ class ProjectSearchResults
...
 
+function get_search_configure_link()
+{
+    $origin = urlencode($_SERVER['REQUEST_URI']);
+    return "<a href='{$_SERVER['PHP_SELF']}?show=config&amp;origin=$origin'>" . _("Configure Result") . "</a>";
+}
+
+function make_link($new_array, $text, $anchor = '')
+{
+    $url = "{$_SERVER['PHP_SELF']}?" . http_build_query(array_replace($_GET, $new_array)) . "$anchor";
```
Comment (cpeel) [inline pinc/ProjectSearchResults.inc]: I realize this was in the code before your change, but we should avoid using `$_SERVER['PHP_SELF']` for security reasons. Odds are good you can just remove it completely and make this `"?" . http_build_query(...`

File: tools/project_manager/projectmgr.php [hunk 3]
```diff
@@ -87,61 +76,50 @@
...
+    // keep "user_all" or we won't know it is user all
+    $_GET = array_merge($_GET, array('project_manager' => $pguser));
     // TRANSLATORS: Abbreviation for Project Manager
     $sub_title = sprintf(_("All PM projects: %s"), $pguser);
 }
 elseif ($show_view == "user_active")
 {
     $condition = "$PROJECT_IS_ACTIVE_sql AND username = '$pguser'";
-    set_session_user_active();
+    $_GET = array_merge($_GET, array('project_manager' => $pguser, 'state' => array_diff($PROJECT_STATES_IN_ORDER, array(PROJ_SUBMIT_PG_POSTED, PROJ_DELETE))));
```
Comment (cpeel) [inline tools/project_manager/projectmgr.php]: Break this across multiple lines, eg:
```
    $_GET = array_merge($_GET, array(
        'project_manager' => $pguser,
        'state' => array_diff($PROJECT_STATES_IN_ORDER, array(PROJ_SUBMIT_PG_POSTED, PROJ_DELETE))
    ));
```

File: tools/project_manager/projectmgr.php [hunk 4]
```diff
@@ -87,61 +76,50 @@
...
-    );
-}
-
-function create_shortcut_link($text, $url, $show_view="")
+function create_shortcut_link($text, $show_val, $show_view="")
 {
-    if($show_view != $url)
-        return "<a href='{$_SERVER['PHP_SELF']}?show=$url'>$text</a>";
+    if($show_view != $show_val)
+        return "<a href='{$_SERVER['PHP_SELF']}?show=$show_val'>$text</a>";
```
Comment (cpeel) [inline tools/project_manager/projectmgr.php]: Again, we should try to remove `$_SERVER['PHP_SELF']` here. I don't think it's needed.

File: tools/search.php [hunk 5]
```diff
@@ -61,15 +55,13 @@
...
+// show must be search
+$condition = $search_form->get_condition();
 
-// use p_search so it will use saved results if needed
-$search_results = new ProjectSearchResults('p_search', "PS");
-echo "<p><a href='{$_SERVER['PHP_SELF']}?show=blank_search_form'>" . _("Start New Search") . "</a> | " . get_refine_search_link() . " | " . $search_results->get_search_configure_link() . "</p>";
+echo "<h1>", _("Search Results"), "</h1>\n";
 
+$search_results = new ProjectSearchResults("PS");
+echo "<p><a href='{$_SERVER['PHP_SELF']}?show=search_form'>" . _("Start New Search") . "</a> | " . get_refine_search_link() . " | " . get_search_configure_link() . "</p>";
```
Comment (cpeel) [inline tools/search.php]: Ditto removing `$_SERVER['PHP_SELF']`

File: pinc/ProjectSearchForm.inc [hunk 6]
```diff
@@ -32,9 +33,9 @@ class ProjectSearchWidget
 
         if ( $this->type == 'text' )
         {
-            if ( isset($search_data[$this->id]) )
+            if ( isset($_GET[$this->id]) )
             {
-                $value_attr = "value='" . attr_safe($search_data[$this->id]) . "'";
+                $value_attr = "value='" . attr_safe($_GET[$this->id]) . "'";
```
Comment (cpeel) [inline pinc/ProjectSearchForm.inc]: This file should `include_once($relPath."misc.inc");` for `attr_safe()` and now `array_get()` and not rely on something else bringing it in.

File: pinc/ProjectSearchForm.inc [hunk 7]
```diff
@@ -334,32 +341,29 @@ class ProjectSearchForm
 
     public function render()
     {
-        if(isset($_SESSION['search_data']))
-            $search_data = $_SESSION['search_data'];
-        else
-            $search_data = array(); // empty array to avoid problems
         echo "<p>" . _("Search for projects matching the following criteria:")."</p>\n";
-        // query data can be too big for GET so use POST
+        // query data can be too big for GET
```
Comment (cpeel) [inline pinc/ProjectSearchForm.inc]: We should either add more information to this comment or remove it.

File: pinc/ProjectSearchForm.inc [hunk 8]
```diff
@@ -334,32 +341,29 @@ class ProjectSearchForm
...
         echo "</table>$table_header";
         while($widget_index < $widget_count)
-            $this->widgets[$widget_index++]->echo_search_item($search_data);
+            $this->widgets[$widget_index++]->echo_search_item();
         echo "
             </table>
             <div class='center-align' style='clear: both;'></div>
             <input type='submit' value='", attr_safe(_("Search")), "'>
-            <input type='button' onclick=\"window.location.assign('{$_SERVER['PHP_SELF']}?show=blank_search_form');\" value='", attr_safe(_("Clear form")), "'>
+            <input type='button' onclick=\"window.location.assign('{$_SERVER['PHP_SELF']}?show=search_form');\" value='", attr_safe(_("Clear form")), "'>
```
Comment (cpeel) [inline pinc/ProjectSearchForm.inc]: Ditto comments below about not using `$_SERVER['PHP_SELF']`

**Comment by 70ray** [comment]
I've tested the cases where $_SERVER['PHP_SELF'] was removed to confirm they all work correctly.

**Pull request merged**